README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# reversetranslate

<!-- badges: start -->

[![Travis build
status](https://travis-ci.org/greg-botwin/reversetranslate.svg?branch=master)](https://travis-ci.org/greg-botwin/reversetranslate)
[![Codecov test
coverage](https://codecov.io/gh/greg-botwin/reversetranslate/branch/master/graph/badge.svg)](https://codecov.io/gh/greg-botwin/reversetranslate?branch=master)
[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![CRAN
status](https://www.r-pkg.org/badges/version/reversetranslate)](https://cran.r-project.org/package=reversetranslate)
<!-- badges: end -->

The goal of `reversetranslate` is to provide a set of R functions to
assist users in reverse translating an amino acid sequence to DNA
nucleotides.

## Installation

You can install the development version from
[GitHub](https://github.com/greg-botwin/reversetranslate) with:

``` r
# install.packages("devtools")
devtools::install_github("greg-botwin/reversetranslate")
```

## General Details

Users can supply their own Codon Frequency Table relevant to the
organism they are most interested in simulating the reverse translation
in or use one of the commonly desired options supplied. Users can also
select between the following three different models of reverse
translation depending on what assumptions you want to make regarding how
redundant codons are chosen and your intended application.

1.  proportional
2.  equal
3.  GC biased

The user may also choose to limit the inclusion of low frequency codons
into the reverse translated nucleotide sequence.

The package currently maintains Codon Frequency Tables for:

  - Homo sapiens
  - Escherichia coli

Other Codon Frequency Tables can be built from your organism of choice
using the `build_codon_tbl` function.

Additional details on package functionality can be found by viewing the
package vignette.

## Example

### Load Package

``` r
library(reversetranslate)
```

### Example Amino Acid Sequence

``` r
minimal_aa_seq
#> [1] "XXXYXXXYYZ"
```

### Example Codon Frequency Table

``` r
minimal_freq_tbl
#>   codon aa prop
#> 1   AAA  X  0.4
#> 2   GGG  X  0.4
#> 3   CCC  X  0.2
#> 4   TTT  Y  1.0
#> 5   GCT  Z  0.2
#> 6   ATG  Z  0.4
#> 7   AAC  Z  0.2
#> 8   CCC  Z  0.2
```

### Example Reverse Translation

``` r
reverse_translate(amino_acid_seq = minimal_aa_seq, codon_tbl = minimal_freq_tbl,
                  limit = 0, model = "proportional")
#> Properly formated Codon Frequency Table
#> [1] "AAACCCGGGTTTAAACCCAAATTTTTTGCT"
```


NEWS.md
# reversetranslate 1.0.0

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(reversetranslate)

test_check("reversetranslate")


tests/testthat/test-check_codon_table.R
context("test-check_codon_table")

test_that("multiplication works", {
  expect_equal(2 * 2, 4)
})


tests/testthat/test-reverse_translate.R
context("test-reverse_translate")

test_that("length output 3 times input", {
  expect_equal(nchar(reverse_translate(amino_acid_seq = minimal_aa_seq, codon_tbl = minimal_freq_tbl)),
                3 * nchar(minimal_aa_seq))
  expect_equal(nchar(reverse_translate(amino_acid_seq = "X", codon_tbl = minimal_freq_tbl)),
               3 * nchar("X"))
})

test_that("lower okay", {
  expect_equal(nchar(reverse_translate(amino_acid_seq = tolower(minimal_aa_seq), codon_tbl = minimal_freq_tbl)),
               3 * nchar(minimal_aa_seq))
})


R/check_limit.R
#' Checks if user supplied amino acid proportion limit threshold is too restrictive.
#'
#' @param codon_tbl  A properly formatted Codon Frequency Table with the following
#'  column names, c("codon", "aa", "prop"). Capitalization of column names should
#'  not matter. The prop column must be of type numeric and all proportions
#'  between 0 and 1. The sum of each codons proportion per amino acid must total
#'  roughly to 1 (0.98 to 1.02 allowed).
#'
#' @param limit A numeric value greater than or equal to 0 and less than or equal
#' to 0.49. Codons with a proportion below specified value will be filtered and
#' will not be avaliable for selection through the reverse translation function.
#'
#' @return A filtered Codon Frequency Table removing all amino acids whose proportion
#' is less than the provided limit.
#' @export
#'
check_limit <- function(codon_tbl, limit) {

  # check limit is of class numeric
  if(is.numeric(limit) != TRUE) {
    stop("Limit must be of type be numeric")
  }

  # check limit between 0 and 0.49
  if(all(limit >= 0 & limit <= 0.49) == FALSE) {
    stop("Limit must be between 0 and 0.49")
  }

  # check to make sure all aa have >= 1 options after limit
  aa_post_filter <- codon_tbl %>%
    dplyr::filter(.data$prop >= limit)

  if(!identical(sort(unique(aa_post_filter$aa)), sort(unique(codon_tbl$aa)))) {
    stop("Codon frequency limit is too high. Not all amino acids have at least 1 remaining codon option.")
  }
  return(aa_post_filter)
}



R/check_codon_table.R
#' Confirm formatting for user supplied Codon Frequency Table.
#'
#' @param codon_tbl A Codon Frequency Table with the following column names,
#'  c("codon", "aa", "prop"). Capitalization of column names should not matter.
#'  The prop column must be of type numeric and all proportions between 0 and 1.
#'  The sum of each codons proportion per amino acid must total roughly to 1
#'  (0.98 to 1.02 allowed).
#'
#' @format codon_tbl (A Codon Frequency Table):
#' \describe{
#'   \item{codon}{Three DNA nucleotides representing the amino acid.}
#'   \item{aa}{One letter abbreviation of the amino acid or * for Stop}
#'   \item{prop}{The proportion the specific codon is present in the genome}
#' }
#' @return An updated Codon Frequency Table with all column names lower case for
#' downstream use.
#'
#' @export
#'
#'
check_codon_table <- function(codon_tbl) {
  colnames(codon_tbl) <- tolower(colnames(codon_tbl))

  # check codon tbl has proper column names
  if(all(c("codon", "aa", "prop") %in% colnames(codon_tbl)) == FALSE) {
    stop("Codon Frequency Table is improperly formated. The table must contain the following
         column names : Codon, Prop, AA")
  }

  # check codon tbl props are of class numeric
  if(is.numeric(codon_tbl$prop) != TRUE) {
    stop("Codon proportions must be numeric")
  }

  # check codon tbl props between 0 and 1
  if(max(codon_tbl$prop) > 1 | min(codon_tbl$prop) < 0) {
    stop("Codon proportions must be between 0 and 1")
  }

  # check the sum of props by aa sum to 0.98 to 1.02
  total_props <- codon_tbl %>%
    dplyr::group_by(.data$aa) %>%
    dplyr::summarise(total_prop = sum(.data$prop)) %>%
    dplyr::select(.data$total_prop) %>%
    unlist()

  if(all(total_props >= 0.98 & total_props <= 1.02) != TRUE) {
    stop("Codon proportions per amino acid, must sum to 1 +/- 0.02")
  }

  else(message("Properly formated Codon Frequency Table"))

  # convert all codons to upper
  codon_tbl$aa <- toupper(codon_tbl$aa)

  return(codon_tbl)

}


R/reverse_translate.R
#' Reverse translate amino acid Sequence to DNA nucleotides
#'
#' @param amino_acid_seq A character vector of representing amino acid(s).
#' @param codon_tbl A properly formatted Codon Frequency Table with the following
#'  column names, c("codon", "aa", "prop"). The Prop column must be of type
#'  numeric and all proportions between 0 and 1. The sum of each codons
#'  proportion per amino acid equal roughly 1 (0.98 to 1.02 allowed). Please see
#'  check_codon_table for more information.
#' @param limit Numerical value set to restrict codons to proportions greater
#' then the set limit. Must be between 0 and 0.49. Defaults to 0, no limit if
#' not set. When limiting rare codons, the residual proportions are split
#' equally amongst the remaining codon options.
#' @param model Either proportional, equal or gc_bias. if a limit is applied,
#' the limit parameter will first be chosen over the GC correction
#'
#' @return character vector representing reverse translated DNA nucleotide codons.
#' @export
reverse_translate <- function(amino_acid_seq, codon_tbl, limit = 0, model = "proportional") {

  # check amino acid sequence is a character string
  if(typeof(amino_acid_seq) != "character") {
    stop("Amino acid sequence must be of type character")
  }

  # convert amino acid sequence to upper
  amino_acid_seq <- toupper(amino_acid_seq)

  # check if single amino_acid_seq or string
  if(nchar(amino_acid_seq) == 1) {
    string <- "single"
  }
  else{string <- "multi"}

  # check codon table
  codon_tbl <- check_codon_table(codon_tbl = codon_tbl)

  # check limit
  check_limit(codon_tbl = codon_tbl, limit = limit)

  # check if model is appropriate
  if(!model %in% c("proportional", "equal", "gc_biased")) {
    stop("Model must be one of the following options: proportional, equal, gc_biased")
  }

  # reverse translate single character
  if(string == "single") {

    # check amino acid in freq table
    if((amino_acid_seq %in% codon_tbl$aa) == FALSE) {
      stop("Amino acid not found in selected codon table")
    }

    # define codon options
    options <- codon_tbl %>%
      dplyr::filter(.data$prop >= limit) %>%
      dplyr::filter(.data$aa == amino_acid_seq)


    result <- select_codon(options = options, model = model)
    return(result)

  }

  # reverse translate multi character
  if(string == "multi") {

    # break apart string into individual amino_acid_seq
    amino_acids <- strsplit(amino_acid_seq, "")[[1]]

    # check all amino_acid_seq in table
    if(!all(unique(amino_acids) %in% unique(codon_tbl$aa))) {
      stop("Not all amino acids found in selected codon table")
    }

    selected_codons <- vapply(amino_acids, function(x) {
      options <- codon_tbl %>%
        dplyr::filter(.data$prop >= limit) %>%
        dplyr::filter(.data$aa == x)

      select_codon(options = options, model = model)
    }, FUN.VALUE = character(1))
    result <- paste0(unlist(selected_codons), collapse = "")
    return(result)
  }
}

select_codon <- function(options, model) {

  # proportional model
  if(model == "proportional") {

    ## calcualte missing prop and add equally
    ## rmultionom distribution based on provided prop
    ## select 1 row

    missing_prop <- (1 - sum(options$prop))

    selected_codon <- options %>%
      dplyr::mutate(updated_prop = .data$prop/sum(options$prop))%>%
      dplyr::mutate(select = c(stats::rmultinom(n = 1, size = 1, prob = .data$updated_prop))) %>%
      dplyr::filter(.data$select == 1) %>%
      dplyr::select(.data$codon) %>%
      unlist() %>%
      as.character()

    return(selected_codon)

  }

  if(model == "equal") {
    ## assign each remaining codon option equal prop
    selected_codon <- options %>%
      dplyr::mutate(updated_prop = 1/nrow(options)) %>%
      dplyr::mutate(select = c(stats::rmultinom(n = 1, size = 1, prob = .data$updated_prop))) %>%
      dplyr::filter(.data$select == 1) %>%
      dplyr::select(.data$codon) %>%
      unlist() %>%
      as.character()

    return(selected_codon)

  }

  if(model == "gc_biased") {
    options <- options %>%
      dplyr::mutate(gc_count = stringr::str_count(.data$codon, "G|C")) %>%
      dplyr::group_by(.data$aa) %>%
      dplyr::mutate(gc_content = dplyr::if_else(.data$gc_count == min(.data$gc_count), "min",
                                         dplyr::if_else(.data$gc_count == max(.data$gc_count), "max", "med"))) %>%
      dplyr::ungroup() %>%
      dplyr::filter(.data$gc_content == "min")

    missing_prop <- (1 - sum(options$prop))

    selected_codon <- options %>%
      dplyr::mutate(updated_prop = .data$prop/sum(options$prop))%>%
      dplyr::mutate(select = c(stats::rmultinom(n = 1, size = 1, prob = .data$updated_prop))) %>%
      dplyr::filter(.data$select == 1) %>%
      dplyr::select(.data$codon) %>%
      unlist() %>%
      as.character()

    return(selected_codon)

  }
}



R/build_hive_codon_tbl.R
#' Converts an organism specific HIVE Codon Usage Table to a Codon Frequency Table
#'
#' For details on the Codon Frequency Table format please see
#' `reversetrasnalte::check_codon_table()`.
#'
#' @param file the name of the file which the data are to be read from
#' @param skip integer, the number of lines of the data file to skip before beginning to read data.
#' @param total_codons total number of codons analysed per organism as reported in HIVE Codon Usage Table
#'
#' @return data.frame
#' @export
#'
build_hive_codon_tbl <- function(file, skip = 2, total_codons){

  # check total_codons numeric
  if(!is.numeric(total_codons)){
    stop("total_codons must be a numeric value equal to the sums of all the counts")
  }

  hive_df <- utils::read.table(file = file, skip = skip)

  hive_df <- hive_df %>%
    tidyr::separate_rows(.data$V1, sep = "\\)") %>%
    dplyr::mutate(codon = stringr::str_extract(.data$V1, "[[:upper:]]{3}")) %>%
    dplyr::mutate(prop_1000 = as.double(stringr::str_extract(.data$V1, "[[:digit:]]+.[[:digit:]]{2}"))) %>%
    dplyr::mutate(count = as.double(stringr::str_extract(.data$V1, "[[:digit:]]{3,}"))) %>%
    dplyr::select(-.data$V1)

  hive_df <- hive_df[stats::complete.cases(hive_df), ]

  count_sum <- sum(hive_df$count)

  if(count_sum != total_codons) {
    stop("Sum of counts does not equal supplied total_cds (Total Coding Sequences (CDS))")
  }

  hive_df <- dplyr::left_join(hive_df, reversetranslate::aa_names, by = c("codon" = "Codon")) %>%
    dplyr::rename("aa" = .data$AA) %>%
    dplyr::group_by(.data$aa) %>%
    dplyr::mutate(aa_total = sum(.data$count)) %>%
    dplyr::ungroup() %>%
    dplyr::mutate(prop = .data$count/.data$aa_total) %>%
    dplyr::select(.data$aa, .data$codon, .data$prop)


  return(hive_df)
}


data-raw/data.R
library(dplyr)

# amino acid naming
aa_names <- utils::read.csv("data-raw/aa_naming.txt", comment.char = "#",
                            header = FALSE, col.names = c("Name", "Abr", "AA", "Codon"),
                            stringsAsFactors = FALSE)

usethis::use_data(aa_names, overwrite = TRUE)

# prepare human codon table HIVE
hsapien_tbl <- reversetranslate::build_hive_codon_tbl("data-raw/hsapien_hive.txt",
                                                           skip = 2, 78580607)

usethis::use_data(hsapien_tbl, overwrite = TRUE)

# prepare E coli table HIVE
ecoli_tbl <- reversetranslate::build_hive_codon_tbl("data-raw/ecoli_hive.txt",
                                                      skip = 2, 17506389367)

usethis::use_data(ecoli_tbl, overwrite = TRUE)

# minimum codon freq table simple tests
minimal_freq_tbl <- data.frame("codon" = c("AAA", "GGG", "CCC", "TTT", "GCT", "ATG", "AAC", "CCC"),
                               "aa" = c("X", "X", "X", "Y", "Z", "Z", "Z", "Z"),
                               "prop" = c(0.4, 0.4, 0.2, 1, 0.2, 0.4, 0.2, 0.2),
                               stringsAsFactors = FALSE)

usethis::use_data(minimal_freq_tbl, overwrite = TRUE)

# mimimum amino acid sequence for simple tests
minimal_aa_seq <- c("XXXYXXXYYZ")

usethis::use_data(minimal_aa_seq, overwrite = TRUE)
